email_verification/models.py

```diff
@@ -0,0 +1,145 @@
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Models for email verification
+"""
+
+# Imports #####################################################################
+
+from datetime import datetime, timedelta, timezone
+from urllib.parse import urljoin
+
+from django.conf import settings
+from django.core.mail import send_mail
+from django.core.urlresolvers import reverse
+from django.db import models
+from django.utils.crypto import get_random_string
+
+
+# Models ######################################################################
+
+class EmailVerification(models.Model):
+    """
+    Verification status for an email address. The main entry point is the
+    `verify` classmethod.
+    """
+    VERIFICATION_CODE_LENGTH = 64
+
+    verification_code_expiry = timedelta(days=7)
+    verification_email_subject = 'Please verify this email address'
+
+    email = models.EmailField(primary_key=True)
+    verified = models.BooleanField(default=False)
+    verification_code = models.CharField(max_length=VERIFICATION_CODE_LENGTH,
+                                         db_index=True)
+    verification_email_date = models.DateTimeField(null=True)
+    verification_date = models.DateTimeField(null=True)
+
+    @classmethod
+    def verify(cls, email, base_url, force_send_email=False):
+        """
+        Verify the given email address. Returns a corresponding
+        `EmailVerification` instance.
+
+        `base_url` should be the domain prefix to include in the verification
+        link, e.g. 'http://example.com/'.
+
+        If an email has already been sent to verify the given address, a new
+        email will not be sent unless `force_send_email` is True.
+        """
+        try:
+            instance = cls.objects.get(email=email)
+        except cls.DoesNotExist:
+            instance = cls(email=email)
+        if force_send_email or not instance.verification_email_sent:
+            instance.send_verification_email(base_url)
+        return instance
+
+    @classmethod
+    def check_verification_code(cls, code):
+        """
+        Check that the given verification code is valid, and mark the email
+        address as verified if it is. Returns the corresponding
+        `EmailVerification` instance.
+        """
+        instance = cls.objects.get(verification_code=code)
```

Comment:
@omarkhan `instance` is a bit generic (and slightly confusing since we also have an `instance` model). Rename to `email_verification`, perhaps?